File: b_rank_new_level_up_problems/b_rank_new_level_up_problems__taxi_distance.cs
// 実行時間 20ms
using static System.Console;
using static System.Linq.Enumerable;
using static System.Math;
using Tpl = System.Tuple<double, int>;

internal class Program
{
    static void Main()
    {
        int[] p = ReadLine().Split().Select( int.Parse ).ToArray();
        int n = int.Parse( ReadLine() );
        Tpl[] d1 = new Tpl[n];
        Tpl[] d2 = new Tpl[n];

        foreach( int i in Range( 0, n ) ) {
            int[] t = ReadLine().Split().Select( int.Parse ).ToArray();

            d1[i] = new Tpl( Sqrt( Pow( p[0] - t[0], 2 ) + Pow( p[1] - t[1], 2 ) ), i + 1 );
            d2[i] = new Tpl( Abs( p[0] - t[0] ) + Abs( p[1] - t[1] ), i + 1 );
        }
        foreach( Tpl i in d1.OrderBy( x => x.Item1 ).ThenBy( x => x.Item2 ).Take( 3 ) ) {
            WriteLine( i.Item2 );
        }
        foreach( Tpl i in d2.OrderBy( x => x.Item1 ).ThenBy( x => x.Item2 ).Take( 3 ) ) {
            WriteLine( i.Item2 );
        }
        return;
    }
}

File: b_rank_new_level_up_problems/b_rank_new_level_up_problems__taxi_distance.py
# 実行時間 20ms
from math import sqrt

px, py = list(map(int, input().split()))
n = int(input())
d1 = [[0, 0] for _ in range(n)]
d2 = [[0, 0] for _ in range(n)]

for i in range(n):
    t = list(map(int, input().split()))

    d1[i][0] = sqrt((px - t[0]) ** 2 + (py - t[1]) ** 2)
    d1[i][1] = i + 1
    d2[i][0] = abs(px - t[0]) + abs(py - t[1])
    d2[i][1] = i + 1
d1.sort(key=lambda x: (x[0], x[1]))
d2.sort(key=lambda x: (x[0], x[1]))
for i in range(3):
    print(d1[i][1])
for i in range(3):
    print(d2[i][1])
